doc/specs/#888 - Com Api.md

```diff
@@ -0,0 +1,889 @@
+make use of that functionality. The API will be preferred by callers that want to receive progress
+and completion events, and UWP packages that do not have permission to launch command line processes.
+The goal for this api is to provide the full set of install functionality possible using the Windows
+Package Manager command line. The command line is documented at
+https://docs.microsoft.com/en-us/windows/package-manager/winget/
+
+# 2. Description
+
+Windows Package Manager is a package manager for windows applications. It comes with a predefined
+repository of applications and users can add new repositories using the winget command line. This
+API allows packaged apps with the packageManagement capability and other higher privilege processes
+to start, manage, and monitor installation of packages that are listed in Windows Package Manager
+repositories.
+
+# 3. Examples
+
+Sample member values for the following examples:
+m_installAppId = L"Microsoft.VSCode";
+
+## 3.1. Create objects
+
+Creation of objects has to be done through CoCreateInstance rather than normal winrt initialization
+since it's hosted by an out of proc com server. These helper methods will be used in the rest of the
+examples.
+
+```c++ (C++ish pseudocode)
+     template<typename TOutput> TOutput ActivateByCoCreate(REFCLSID rclsid) {
+        winrt::com_ptr<::IInspectable> result;
+        check_hresult(::CoCreateInstance(rclsid, nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(&result)));
+        return result.as<TOutput>();
+    }
+
+    AppInstaller CreateAppInstaller() {
+        return ActivateByCoCreate<AppInstaller>(CLSID_AppInstaller);
+    }
+    InstallOptions CreateInstallOptions() {
+        return ActivateByCoCreate<InstallOptions>(CLSID_InstallOptions);
+    }
+    FindPackagesOptions CreateFindPackagesOptions() {
+        return ActivateByCoCreate<FindPackagesOptions>(CLSID_FindPackagesOptions);
+    }
+    GetCompositeAppCatalogOptions CreateGetCompositeAppCatalogOptions() {
+        return ActivateByCoCreate<GetCompositeAppCatalogOptions>(CLSID_GetCompositeAppCatalogOptions);
+    }
+    PackageMatchFilter CreatePackageMatchFilter() {
+        return ActivateByCoCreate<PackageMatchFilter>(CLSID_PackageMatchFilter);
+    }
+```
+
+## 3.1. Search
+
+The api can be used to search for packages in a catalog known to Windows Package Manager. This can
+be used to get availability information or start an install.
+
+```c++ (C++ish pseudocode)
+
+    IAsyncOperation<CatalogPackage> FindPackageInCatalog(AppCatalog catalog, std::wstring packageId)
+    {
+        FindPackagesOptions findPackagesOptions = CreateFindPackagesOptions();
+        PackageMatchFilter filter = CreatePackageMatchFilter();
+        filter.IsAdditive(true);
+        filter.Field(PackageMatchField::Id);
+        filter.Type(MatchType::Exact);
+        filter.Value(packageId);
+        findPackagesOptions.Filters().Append(filter);

```

Comment:
What happens if multiple filters are appended? I see the documentation for FindPackageOptions says that only one filter may be applied at a time.

Reply:
It says only one filter with match type AppCatalogDefined. Multiple filters of other match types are allowed. I am not totally sure of the history behind that implementation limitation or how hard it would be to change it. At some point there is likely a limit to how complex we want to allow queries to get.

Reply:
Ahh.  Yes.  I did misread the comment.  My mistake.